### ISSUE TITLE
why the Chebyshev polynomials is needed?

### ISSUE THREAD

**Issue by zeal-up** [mdeff/cnn_graph#35]
When approximate the initial filters kernel gθ(Λ) = diag(θ) with a polynomial 
![image](https://user-images.githubusercontent.com/40826585/44521567-1512dc00-a706-11e8-8219-2b85f0f71002.png)
why we can‘t simply combine the eigenvector and eigenvalue matrix together like
![image](https://user-images.githubusercontent.com/40826585/44521653-5efbc200-a706-11e8-8920-4dcd9c3e2c5d.png)
because 
![image](https://user-images.githubusercontent.com/40826585/44521667-6ae78400-a706-11e8-89f8-af9c3b2a941a.png) *
The paper instead, use Chebyshev polynomials and also said that 
![image](https://user-images.githubusercontent.com/40826585/44521762-c6197680-a706-11e8-95a9-ac5086a95c44.png)
I'm quite confused here. I don't know why the Chebyshev is needed?

**Comment by ZiqianXie** [comment]
+1, I was doing GCN paper review and read this paper today, came up with the same question.
We can compute Lx, L^2x, ..., L^Kx sequentially using O(K\epsilon) operations, and linearly combine them, same computational complexity.

**Comment by ZiqianXie** [comment]
I think in the cited wavelet on graph paper (Hammond et al.), they used chebyshev polynomial to approximate a known wavelet function because chebyshev interpolation has certain desired property, in the case where the filters are unknown, the chebyshev polynomials are not necessary, any linear combination of chebyshev polynomials up to Kth degree can be represented by a Kth degree polynomial.

**Comment by u42dani** [comment]
Has anyone tried node classification on a general graph without chebyshev polynomials (i.e. using only the laplacian and its powers)? What was the accuracy compared to the chebyshev method?

**Comment by Keramatfar** [comment]
See here:
https://www.youtube.com/watch?v=v3jZRkvIOIM.
min 26.

**Comment by Jovonni** [comment]
> Has anyone tried node classification on a general graph without chebyshev polynomials (i.e. using only the laplacian and its powers)? What was the accuracy compared to the chebyshev method?

I have used the laplacian, and the normalized laplacian to train a convolution NN. The accuracy on the training, testing, and validation set were all above 85%.

The data I used was proxy related data and the model was for a security related to problem. However, doing that is how I found this work, and xaviers work as a whole... so I’ve seen it work, but Xavier Bresson and his team build on those concepts so well....

So, not the best modeling approach, but it is tractable for certain problems.

**Comment by ZiqianXie** [comment]
> 
> 
> See here:
> https://www.youtube.com/watch?v=v3jZRkvIOIM.
> min 26.

Yes, it is stable under coefficient perturbation but otherwise the complexities are the same.

**Comment by mdeff** [comment]
Any polynomial (monomials, Chebyshev, Legendre, etc.) of order K has the same representation power and can be used. We used Chebyshev because we had experience with it and thought it would be easier to optimize as they form an orthogonal basis. There's however not much difference in practice.

Chebyshev polynomials are more orthogonal (under the identity measure, and [truly orthogonal with measure `1/sqrt(1-x²)`](https://en.wikipedia.org/wiki/Chebyshev_polynomials#Orthogonality)) than monomials, but less than Legendre ([truly orthogonal with the identity measure](https://en.wikipedia.org/wiki/Legendre_polynomials#Orthonormality_and_completeness)):
![polynomial_bases_spectrum](https://user-images.githubusercontent.com/6806065/87963678-632ee680-cab9-11ea-9784-cc6284268772.png)

In the vertex domain (for a ring/circle graph), Chebyshev and Legendre polynomials are also more orthogonal than monomials:
![polynomial_bases_vertex](https://user-images.githubusercontent.com/6806065/87963673-61fdb980-cab9-11ea-807b-8f1e7344c2ec.png)

**Comment by youkaichao** [comment]
I have the same question as @zeal-github , and opened a question in [stack exchange](https://stats.stackexchange.com/q/500392/215724). After reading this issue, I think the question is clear now. Chebyshev polynomials may have better property with respect to parameter orthogonality, but the computation complexity is the same as directly using laplacian matrix.

**Comment by hazdzz** [comment]
> Any polynomial (monomials, Chebyshev, Legendre, etc.) of order K has the same representation power and can be used. We used Chebyshev because we had experience with it and thought it would be easier to optimize as they form an orthogonal basis. There's however not much difference in practice.
> 
> 
> 
> Chebyshev polynomials are more orthogonal (under the identity measure, and [truly orthogonal with measure `1/sqrt(1-x²)`](https://en.wikipedia.org/wiki/Chebyshev_polynomials#Orthogonality)) than monomials, but less than Legendre ([truly orthogonal with the identity measure](https://en.wikipedia.org/wiki/Legendre_polynomials#Orthonormality_and_completeness)):
> 
> ![polynomial_bases_spectrum](https://user-images.githubusercontent.com/6806065/87963678-632ee680-cab9-11ea-9784-cc6284268772.png)
> 
> 
> 
> In the vertex domain (for a ring/circle graph), Chebyshev and Legendre polynomials are also more orthogonal than monomials:
> 
> ![polynomial_bases_vertex](https://user-images.githubusercontent.com/6806065/87963673-61fdb980-cab9-11ea-807b-8f1e7344c2ec.png)
> 
> 

Why did you choose the Chebyshev Polynomials of first kind instead of the Legendre Polynomials?

**Comment by mdeff** [comment]
Mostly for historical reasons. At first we were *designing* filters (e.g., to solve the diffusion of heat, the propagation of waves, and [many others](https://pygsp.readthedocs.io/en/latest/reference/filters.html)). As Chebyshev polynomials are excellent [function approximators](https://en.wikipedia.org/wiki/Approximation_theory#Chebyshev_approximation), we chose them to approximate those ideal filters we wanted to design. When *learning* filters, the choice of a polynomial basis doesn't matter for the expressivity of the filters, so we kept Chebyshev as we were familiar with them. We thought it could be easier to optimize, but that doesn't seem to make a difference in practice.

**Comment by hazdzz** [comment]
> Mostly for historical reasons. At first we were *designing* filters (e.g., to solve the diffusion of heat, the propagation of waves, and [many others](https://pygsp.readthedocs.io/en/latest/reference/filters.html)). As Chebyshev polynomials are excellent [function approximators](https://en.wikipedia.org/wiki/Approximation_theory#Chebyshev_approximation), we chose them to approximate those ideal filters we wanted to design. When *learning* filters, the choice of a polynomial basis doesn't matter for the expressivity of the filters, so we kept Chebyshev as we were familiar with them. We thought it could be easier to optimize, but that doesn't seem to make a difference in practice.

In your opinions, it doesn't matter which orthogonal polynomials are chosen. Because in practice, the performance is not depends on the kinds of orthogonal polynomials. Am I understanding correctly?

**Comment by mdeff** [comment]
Yes. It seems that SGD works well enough to find the best polynomial, whatever the basis. That might also be because we use low-order polynomials when learning.

**Comment by hazdzz** [comment]
> Yes. It seems that SGD works well enough to find the best polynomial, whatever the basis. That might also be because we use low-order polynomials when learning.

What is SGD?

**Comment by mdeff** [comment]
[Stochastic Gradient Descent](https://en.wikipedia.org/wiki/Stochastic_gradient_descent)

**Comment by hazdzz** [comment]
> Mostly for historical reasons. At first we were *designing* filters (e.g., to solve the diffusion of heat, the propagation of waves, and [many others](https://pygsp.readthedocs.io/en/latest/reference/filters.html)). As Chebyshev polynomials are excellent [function approximators](https://en.wikipedia.org/wiki/Approximation_theory#Chebyshev_approximation), we chose them to approximate those ideal filters we wanted to design. When *learning* filters, the choice of a polynomial basis doesn't matter for the expressivity of the filters, so we kept Chebyshev as we were familiar with them. We thought it could be easier to optimize, but that doesn't seem to make a difference in practice.

I have another big problem. Does the orthogonal polynomials causes low-pass filter? Or Spectral CNN (_Spectral Networks and Deep Locally Connected Networks on Graphs_) itself is a low-pass filter GNN?

**Comment by mdeff** [comment]
Neither spectral nor polynomial convolutions enforce low-pass filters. Polynomial convolutions enforce local filters (the locality is controlled by the order of the polynomials). Local filters can be high-pass (simply think about a multiplication by the Laplacian itself, a second-order derivative). The low-pass limitation is specific to GCN [Kipf et al.] as they merged the 0- and 1-neighborhood in the quest of learning a single parameter per filter.

**Comment by hazdzz** [comment]
> more orthogonal

What do you mean more orthogonal?

**Comment by hazdzz** [comment]
> Any polynomial (monomials, Chebyshev, Legendre, etc.) of order K has the same representation power and can be used. We used Chebyshev because we had experience with it and thought it would be easier to optimize as they form an orthogonal basis. There's however not much difference in practice.
> 
> Chebyshev polynomials are more orthogonal (under the identity measure, and [truly orthogonal with measure `1/sqrt(1-x²)`](https://en.wikipedia.org/wiki/Chebyshev_polynomials#Orthogonality)) than monomials, but less than Legendre ([truly orthogonal with the identity measure](https://en.wikipedia.org/wiki/Legendre_polynomials#Orthonormality_and_completeness)):
> ![polynomial_bases_spectrum](https://user-images.githubusercontent.com/6806065/87963678-632ee680-cab9-11ea-9784-cc6284268772.png)
> 
> In the vertex domain (for a ring/circle graph), Chebyshev and Legendre polynomials are also more orthogonal than monomials:
> ![polynomial_bases_vertex](https://user-images.githubusercontent.com/6806065/87963673-61fdb980-cab9-11ea-807b-8f1e7344c2ec.png)

What do you mean more orthogonal?
